plugin/trino-phoenix5/src/main/java/io/trino/plugin/phoenix5/PhoenixMetadata.java

```diff
@@ -273,6 +282,52 @@ public void dropTable(ConnectorSession session, ConnectorTableHandle tableHandle
         phoenixClient.dropTable(session, (JdbcTableHandle) tableHandle);
     }
 
+    @Override
+    public RowChangeParadigm getRowChangeParadigm(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        return CHANGE_ONLY_UPDATED_COLUMNS;
+    }
+
+    @Override
+    public JdbcColumnHandle getMergeRowIdColumnHandle(ConnectorSession session, ConnectorTableHandle tableHandle)
+    {
+        JdbcTableHandle handle = (JdbcTableHandle) tableHandle;
+
+        ImmutableList<RowType.Field> fields = phoenixClient.getPrimaryKeyColumnHandles(session, handle).stream()
+                .map(columnHandle -> new RowType.Field(Optional.of(columnHandle.getColumnName()), columnHandle.getColumnType()))
+                .collect(toImmutableList());
+        verify(!fields.isEmpty(), "Phoenix primary key is empty");
+
+        return new JdbcColumnHandle(
+                MERGE_ROW_ID_COLUMN_NAME,
+                new JdbcTypeHandle(Types.ROWID, Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()),
+                RowType.from(fields));
+    }
+
+    @Override
+    public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode)
+    {
+        JdbcTableHandle handle = (JdbcTableHandle) tableHandle;
+        checkArgument(handle.isNamedRelation(), "Merge target must be named relation table");
+        JdbcTableHandle plainTable = phoenixClient.buildPlainTable(handle);
+        JdbcColumnHandle mergeRowIdColumnHandle = getMergeRowIdColumnHandle(session, plainTable);
+
+        List<JdbcColumnHandle> columns = phoenixClient.getColumns(session, plainTable).stream()
+                .filter(column -> !ROWKEY.equalsIgnoreCase(column.getColumnName()))
+                .collect(toImmutableList());
+        PhoenixOutputTableHandle phoenixOutputTableHandle = (PhoenixOutputTableHandle) beginInsert(session, plainTable, ImmutableList.copyOf(columns), retryMode);
+
+        return new PhoenixMergeTableHandle(
+                phoenixClient.updatedScanColumnTable(session, handle, handle.getColumns(), mergeRowIdColumnHandle),
+                phoenixOutputTableHandle,
+                mergeRowIdColumnHandle);
+    }
+
+    @Override
+    public void finishMerge(ConnectorSession session, ConnectorMergeTableHandle tableHandle, Collection<Slice> fragments, Collection<ComputedStatistics> computedStatistics)
```

Comment:
AIUI the body of `finishMerge` can be empty because `storeMergedRows` does everything needed to finish the operation.

I do wonder about the `computedStatistics`.  Do Jdbc connectors always ignore statistics?  The main thing non-Jdbc connectors get from the statistics is an accurate row count.